### PR TITLE
Bug 1710502: Using trap to deal with signaling on node-ca.

### PR DESCRIPTION
**Pull request by ricardomaraschini**
This patch adds a trap to deal with SIGTERM, allows sleep to be
interrupted by signaling.

### CODE REVIEW

**Comment by ricardomaraschini** [comment]
> @ricardomaraschini note - per conversation with @dmage you can't "scale down" a Daemonset. There's two options:
> 
> 1. Delete the daemonset
> 2. Pick a pod managed by the daemonset. Delete it, wait one minute, then verify if:
>    a. The pod does not exist or,
>    b. The node-ca container is Terminated with exit code 0

One minute is way too much as the pod will probably receive a SIGKILL after 30 seconds, dying miserably.

**Comment by ricardomaraschini** [comment]
> @ricardomaraschini note - per conversation with @dmage you can't "scale down" a Daemonset. There's two options:
> 
> 1. Delete the daemonset
> 2. Pick a pod managed by the daemonset. Delete it, wait one minute, then verify if:
>    a. The pod does not exist or,
>    b. The node-ca container is Terminated with exit code 0

Done. Could you take a look?

**Comment by ricardomaraschini** [comment]
/hold

Tests on this one are failing while we wait for https://github.com/openshift/image-registry/pull/192
After that one gets merged we can rebase this one and everything should work.

File: test/framework/logs.go [hunk 1]
```diff
@@ -97,3 +98,37 @@ func DumpPodLogs(logger Logger, podLogs PodSetLogs) {
...
+		defer close(out)
+
+		r := bufio.NewReader(ls)
+		for {
+			line, err := r.ReadString('\n')
+			switch {
+			case err == io.EOF:
+				return
+			case err != nil:
+				t.Fatal("error reading line", err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply test/framework/logs.go]: Are you saying that it does not work or that it is not the proper way to do it?

File: test/framework/imageregistry.go [hunk 2]
```diff
@@ -187,6 +187,28 @@ func DeployImageRegistry(logger Logger, client *Clientset, cr *imageregistryapiv
 	return nil
 }
 
+// MustDeployImageRegistryWithConfig deploys image registry using a default
+// config reference. It is a helper for MustDeployImageRegistry().
+func MustDeployImageRegistryWithConfig(t *testing.T, client *Clientset) {
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply test/framework/imageregistry.go]: Done.

File: test/framework/logs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply test/framework/logs.go]: Nevermind, I understood what you mean. Thanks for catching this!

[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply test/framework/logs.go]: @dmage could you check if it is OK now?

File: test/e2e/graceful_shutdown_test.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	); err != nil {
+		t.Fatalf("error deleting pod: %v", err)
+	}
+
+	timeout := time.NewTimer(time.Minute)
+	defer timeout.Stop()
+	for {
+		select {
+		case <-timeout.C:
+			t.Fatal("timeout awaiting for pod to die.")
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply test/e2e/graceful_shutdown_test.go]: Could you elaborate more? My reasoning for this is that if it is taking more than one minute for the pod to die there must be something very wrong. After this period of time the pod would have already received a SIGKILL. At this is a critical condition I have used Fatal.

File: test/e2e/graceful_shutdown_test.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+			t.Fatalf("error reading pod log: %v", err)
+		case line, more := <-logch:
+			if !more {
+				t.Fatal("pod died, no graceful message found.")
+			}
+			// this is the log line image registry pod prints when
+			// it gracefuly dies, if we find this it means that the
+			// pod had manage to exit properly so we can end this
+			// test successfuly.
+			if strings.HasPrefix(line, "server shutdown, bye.") {
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply test/e2e/graceful_shutdown_test.go]: I am checking this.

File: test/e2e/graceful_shutdown_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply test/e2e/graceful_shutdown_test.go]: > Fatal means end the test now, Error means the test will fail, but other checks can run.
> 
> I tend to use `Fatal` if a precondition for the test isn't satisfied, `Error` on things I'm verifying. In this case we are trying to verify that the pod terminates in under a minute.

I completely agree with your reasoning around the preconditions and i will use it from now on on my future PRs. 

What we are testing here is not if the pod terminates in under one minute but if it prints a specific message while dying(gracefully). I would consider the "one minute timeout" as a precondition for the test. We are giving the pod one minute to die, if it does not die then we fail.

**Comment by smarterclayton** [comment]
Is there anything left on this?

**Comment by ricardomaraschini** [comment]
> Is there anything left on this?

We were having problems with testings until yesterday but now it is just a matter of approvals/lgtm.

**Pull request merged**